ocaml/typing/typecore.ml

```diff
@@ -415,50 +406,90 @@ let check_tail_call_local_returning loc env ap_mode {region_mode; _} =
        ap_mode is local, the application allocates in the outer
        region, and thus [region_mode] needs to be marked local as well*)
       match
-        Regionality.submode (Regionality.of_locality ap_mode) region_mode
+        Regionality.submode (locality_as_regionality ap_mode) region_mode
       with
       | Ok () -> ()
       | Error _ -> raise (Error (loc, env, Tail_call_local_returning))
     end
   | None -> ()
 
+let meet_regional mode =
+  let mode = Value.disallow_left mode in
+  Value.meet [mode; (Value.max_with_regionality Regionality.regional)]
+
+let meet_global mode =
+  (* let mode = Value.disallow_left mode in *)
+  Value.meet [mode; (Value.max_with_regionality Regionality.global)]
+
+let meet_unique mode =
+  Value.meet [mode; (Value.max_with_uniqueness Uniqueness.unique)]
+
+let meet_many mode =
+  Value.meet [mode; (Value.max_with_linearity Linearity.many)]
+
+let join_shared mode =
+  Value.join [mode; Value.min_with_uniqueness Uniqueness.shared]
+
+let value_regional_to_local mode =
+  mode
+  |> value_to_alloc_r2l
+  |> alloc_as_value
+
+let value_regional_to_global mode =
+  mode
+  |> value_to_alloc_r2g
+  |> alloc_as_value
+
 (* Describes how a modality affects field projection. Returns the mode
    of the projection given the mode of the record. *)
 let _modality_unbox_left global_flag mode =
   match global_flag with
   | Global ->
-      mode |> Value.to_global |> Value.to_shared |> Value.to_many
+      mode
+      |> Value.set_regionality_min
```

Comment:
It's like this because `set_uniqueness_max` doesn't have a total right adjoint, so can't be used on the left. Although, maybe now the solver is ok with partial right adjoints being used on the left? Since I think maybe supporting numbered regions required that. It might still be better to use `join` though because that is obviously correct.

Reply:
(As a side-note, I'm not sure I like the names of the form `set_regionality_min` because they aren't really setting the regionality to min, they are making min an upper bound. In particular, for modes that are neither the min nor the max, I think they have two different morphisms, one to make it an upper bound and one to make it a lower bound.)

Reply:
But as far as I can tell, `join_shared` and `set_uniqueness_max` have the same net effect. The former does it by using a bound, while the latter does it by just creating a constant uniqueness mode, but I think these will behave the same in the submoding functions. I hear what you're saying about adjoints, but if you're saying this can't be done on the left and yet we're doing this on the left, then something's funny somewhere.

Reply:
Sidenote: Currently uniqueness is the only axis in the modality fragment, and therefore `set_uniqueness_min` `set_uniqueness_max` can all be defined without any restriction on the left/right status. However, that's an implementation detail that will soon be invalidated (as we add more axes to the modality fragment) and not exposed.

Reply:
Just to clarify, the "extra lines of code" has been implemented - it's for the finer regionality PR but backported to this PR. 
It's the "implicit injection"/"optimisation vs soundness" comment that we discussed, Richard.

Reply:
> A simple version of our solver needs every morphism used on the left to have a total right adjoint. Separately from that it also understands how to handle meets and joins. Joining with a constant does not necessarily have a total right adjoint, but all joins can instead be handled in the inference via a diffierent approach.

Ah... this makes some sense. If I can reformulate (in the context of the simple version of the the solver): Types aside, `join_shared` and `set_uniqueness_max` are observationally equivalent. However, they manipulate the internal representations of modes in different ways. And furthermore, `set_uniqueness_max` manipulates the internal representation in such a way that the solver would choke if it saw the result of that function on the left. And so because we sometimes want to perform this operation on a mode that will be used on the left, we must use the less-efficient `join_shared`.

That all makes sense to me, and also demonstrates that the problem isn't just that we're being picky about types, but that there is an actual important functional difference between these two (though one not observable outside of the solver implementation).

Reply:
If the extra line of code has been added, though, then can we leverage that extra functionality to unify these functions? (At my most recent pass-through, I realized I didn't understand the "implicit injection" comment but decided not to push. How does implicit injection arise?)

Reply:
Yes, your reformulate sounds correct to me. Although, note that `join_shared` and `set_uniqueness_max` will stop being "somewhat equivalent", if we add a even higher mode than `shared`. They are really just different. 

Also, I'm not sure if `join_shared` is less efficient - I haven't done any profile, but I would say they are both fast.

Reply:
About the "extra lines of code": For the finer regionality PR, we need the solver to be able to handle "partial" adjunction. When the solver is requested `f x <= a`, it can't just deduce `x <= f' a` where `f'` is the right adjoint. Instead, the `f'` given by the user might be a partial morphism and not defined on `a`. What is guaranteed, tho, is that "if `a` is in `f`'s codomain, then `f' a` will be well-defined and `x <= f' a`".  The comment in the solver about "implicit injection" is talking about this.

As I revisit the finer regionality PR, I realize that we use explicit injection in the end, but we still need partial adjunction for some reason. See the `Inj_l` morphism in the finer regionality PR. I will update the comment to reflect that.

Reply:
I've made internal ticket 2297 to continue thinking about this.